lib/src/new-select/NewSelect.jsx

```diff
@@ -0,0 +1,699 @@
+        changeVisualFocusedSuggIndex(-1);
+        if (options[0].options) {
+          setFilteredOptions(
+            options.map((optionGroup) => {
+              let group = JSON.parse(JSON.stringify(optionGroup)); // anything better?
+              group.options = group.options.filter((option) =>
+                option.label.toUpperCase().includes(searchValue.toUpperCase())
+              );
+              return group;
+            })
+          );
+        } else
+          setFilteredOptions(
+            options.filter((option) => option.label.toUpperCase().includes(searchValue.toUpperCase()))
+          );
+      }
+    }, [options, searchable, searchValue]);
+
+    const lastIndex = (isGroupedOptions) => {
+      let last = 0;
+
+      if (isGroupedOptions) {
+        filteredOptions.length
+          ? filteredOptions.forEach((option) => (last += option.options.length))
+          : options.forEach((option) => (last += option.options.length));
+      } else last = filteredOptions.length ? filteredOptions.length - 1 : options.length - 1;
+
+      return last;
+    };
+    const Option = ({ option, index, isGroupedOption = false }) => {
+      const isSelected = (value ?? innerValue) === option.value;
+      const isLastOption = index === lastIndex(isGroupedOption);
+
+      return (
+        <OptionItem
+          onMouseDown={(event) => {
+            event.button === 0 && changeIsActiveOption(true); // left button only
+          }}
+          onMouseUp={(event) => {
+            if (event.button === 0 && isActiveOption) {
+              // left button only
+              handleSelectChangeValue(option.value);
+              setSearchValue("");
+              changeIsActiveOption(false);
+              closeOptions(); // if not multiple
+            }
+          }}
+          onMouseEnter={() => {
+            changeVisualFocusedSuggIndex(index);
+          }}
+          onMouseLeave={() => {
+            changeIsActiveOption(false);
+          }}
+          visualFocused={visualFocusedSuggIndex === index}
+          active={visualFocusedSuggIndex === index && isActiveOption}
+          selected={isSelected}
+        >
+          <StyledOption
+            visualFocused={visualFocusedSuggIndex === index}
+            active={visualFocusedSuggIndex === index && isActiveOption}
+            selected={isSelected}
+            last={isLastOption}
+          >
+            <OptionLabel grouped={isGroupedOption}>{option.label}</OptionLabel>
+            {isSelected && <SelectedIcon>{selectIcons.selected}</SelectedIcon>}
+          </StyledOption>
+        </OptionItem>
+      );
+    };
+    let global_index = 0;
+    const mapOptionFunc = (option, index) => {
+      if (option.options) {
+        return (
+          <>
+            {option.options.length > 0 && <OptionGroupLabel>{option.label}</OptionGroupLabel>}
+            {option.options.map((singleOption) => {
+              global_index++;
```

Comment:
The second param of the map function is the index of the current element. If you do `option.options.map((singleOption), i) => ...`, you can use it to pass it to the `Option` like `index={i}` and you don't need to create a counter (`global_index`).

Reply:
The main goal of using a global variable instead of the regular map index is that it preserves its value between `mapOptionFunc` calls. What i'm trying to achieve is that every option of each group has its own index, to be able for visual focus to identify which is been hovered when the mouse enters and paint it properly. If i use the index of map, it will always start with 0, causing this to happen: 

![image](https://user-images.githubusercontent.com/44321109/144282105-78bbba4b-ec32-4aee-8827-88eb5e223212.png)